### DISCUSSION TITLE
Nondeterministic error with server/client code sharing

### DISCUSSION THREAD

**Discussion by btslater** [remix-run/remix#4671] [Q&A]
### BACKGROUND:

I have a simple app that relies on a proprietary 3rd-party library (which I will call `ThirdParty.js`) which uses browser APIs, and I believe has other side-effects that preclude its use on the server. I'm trying to integrate this into a Remix app. Because of the browser APIs/side effects, I am structuring my code so that the component that uses this 3rd party library is _only_ imported on the client, not the server. Everything works as expected when I run my app locally.

### PROBLEM:

When deployed on a server, _approximately one-seventh of the time_, the server apparently tries importing `ThirdParty.js` code and throws errors; in the browser dev tools, these errors reference obfuscated 3rd-party code that tries to access `window` (a browser-only API). This is surprising to me because I have taken steps to ensure the import of this problematic code should only happen on the client.

These errors appear to be non-deterministic: sometimes the app works as expected, other times it crashes with a `routeModules` error, suggesting an issue with sharing code between client/server. In my hands, the crashes happen only in incognito mode, although my colleagues see crashes in non-incognito.

These errors ONLY appear when running a production build remotely—never a local dev build.

### I HAVE TRIED:

- Putting all code which imports from `ThirdParty.js` inside a `ThirdPartyWrapper.client.jsx` file, 
- this wrapped component is `lazy`-loaded, 
- wrapped in a `Suspense`, 
- AND wrapped in a `ClientOnly`
- … but still the browser errors with: `SyntaxError: Invalid or unexpected token at ThirdPartyWrapper.client-asdfdasjs:22:2691` (which references 3rd-party obfuscated code trying to access `window`). I interpret this to mean that the code is STILL running on the server somehow (maybe something is wrong with the bundling process?)

Is is possible that Remix is somehow aggressively bundling this code and running it on the server, despite my layered attempts to prevent this? The inconsistency/non-repeatability makes me wonder if there could be some kind of race condition, or (module?) cache issue?

Do you all have any idea what might be going wrong leading to this problem, or any advice? I wish I could upload specific steps to repeat, but I would be breaking the TOS for the ThirdParty.js library...

**Comment by omakise** [comment]
same question

**Comment by machour** [comment]
Can you setup a repository reproducing the problem?
